pkg/pq/priority_queue.go

```diff
@@ -51,16 +51,31 @@ func (pq *PriorityQueue) Len() int {
 }
 
 func (pq *PriorityQueue) Release(value Value) {
-	queue := &internalQueue{}
-	heap.Init(queue)
+	idx := pq.findIdx(func(item *Item) bool {
+		return item.value == value
+	})
 
-	for _, item := range *pq.queue {
-		if item.value != value {
-			heap.Push(queue, item)
-		}
+	if idx == -1 {
+		return
+	}
+
+	lastItem := pq.queue.Pop().(*Item)
+
+	if (*pq.queue).Len() == 0 || idx == (*pq.queue).Len() {
+		return
 	}
 
-	pq.queue = queue
+	(*pq.queue)[idx] = lastItem
+	heap.Fix(pq.queue, idx)
+}
+
+func (pq *PriorityQueue) findIdx(fn func(*Item) bool) int {
+	for i, item := range *pq.queue {
+		if fn(item) {
+			return i
+		}
+	}
+	return -1
```

Comment:
I think we abstracted it too quickly since there is only one place to use `findIdx` yet. How about inline it now and extract it later when duplicates occur?